infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -454,6 +461,24 @@ public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
       throws InfrastructureException {
     String namespace = evalPlaceholders(defaultNamespaceName, resolutionCtx);
 
+    if (!NamespaceNameValidator.isValid(namespace)) {
+      Optional<KubernetesNamespaceMeta> namespaceMetaOptional;
+      String normalizedNamespace = NamespaceNameValidator.normalize(namespace);
+      if (normalizedNamespace.isEmpty()) {
+        throw new InfrastructureException(
+            format(
+                "Evaluated empty namespace name for workspace %s", resolutionCtx.getWorkspaceId()));
+      }
+      do {
+        normalizedNamespace =
+            normalizedNamespace
+                .substring(0, Math.min(55, normalizedNamespace.length()))
+                .concat(NameGenerator.generate("-", 6));
+        namespaceMetaOptional = fetchNamespace(normalizedNamespace);
+      } while (namespaceMetaOptional.isPresent());
```

Comment:
can we please limit this loop to some finite number of tries and fail in case it runs out? It's a bit scary like this.

Reply:
I'm more worried about our failures, like `fetchNamespace` method is wrong (e.g. some wrong refactoring), or some bug in `NameGenerator`, or even here with modyfiing the namespace name to test...